rpcs3/Emu/Cell/Modules/cellRtc.cpp

```diff
@@ -517,70 +694,132 @@ error_code cellRtcParseDateTime(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDat
-	else
+	else if (std::isdigit(pszDateTime[pos]))
 	{
-		iss >> std::get_time(&t, "%Y-%m-%dT%H:%M:%S") >> tz; // rfc3339 2020-04-03T13:23:30.30Z
-		if (!iss.fail())
+		day = digit(pszDateTime[pos]);
+
+		if (std::isdigit(pszDateTime[pos + 1]))
 		{
-			// TODO timezone
+			day = day * 10 + (digit(pszDateTime[pos + 1]));
+
+			pos += 2;
 		}
 		else
 		{
-			// TODO asctime
-			iss >> std::get_time(&t, "%a %b  %d %H:%M:%S %Y");//Mon Apr  6 21:58:35 2020
-			if (!iss.fail())
-			{
-				// TODO timezone
-			}
-			else
-			{
-				return CELL_RTC_ERROR_BAD_PARSE;
-			}
+			pos++;
 		}
 	}
+	else
+	{
+		return { CELL_RTC_ERROR_BAD_PARSE, "cellRtcParseDateTime(): failed to parse day: ASCII value 0x%x at position %d is not a digit or space", pszDateTime[pos], pos };
+	}
+
+	// Hour: " X" or " XX"
+	const u16 hour = rtcParseComponent(pszDateTime, pos, ' ', "hour");
+
+	if (hour == umax)
+	{
+		return CELL_RTC_ERROR_BAD_PARSE;
+	}
+
+	// Minute: ":X" or ":XX"
+	const u16 minute = rtcParseComponent(pszDateTime, pos, ':', "minute");
+
+	if (minute == umax)
+	{
+		return CELL_RTC_ERROR_BAD_PARSE;
+	}
+
+	// Second: ":X" or ":XX"
+	const u16 second = rtcParseComponent(pszDateTime, pos, ':', "second");
+
+	if (second == umax)
+	{
+		return CELL_RTC_ERROR_BAD_PARSE;
+	}
+
+	// Mandatory space
+	if (pszDateTime[pos] != ' ')
+	{
+		return { CELL_RTC_ERROR_BAD_PARSE, "cellRtcParseDateTime(): no space after second" };
+	}
+
+	pos++;
+
+	// Year: XXXX
+	if (!std::isdigit(pszDateTime[pos]) ||
+		!std::isdigit(pszDateTime[pos + 1]) ||
+		!std::isdigit(pszDateTime[pos + 2]) ||
+		!std::isdigit(pszDateTime[pos + 3]))
+	{
+		return { CELL_RTC_ERROR_BAD_PARSE, "cellRtcParseDateTime(): failed to parse year: one of the ASCII values 0x%x, 0x%x, 0x%x, or 0x%x is not a digit",
+			pszDateTime[pos], pszDateTime[pos + 1], pszDateTime[pos + 2], pszDateTime[pos + 3] };
+	}
+
+	const u16 year = digit(pszDateTime[pos]) * 1000 + digit(pszDateTime[pos + 1]) * 100 + digit(pszDateTime[pos + 2]) * 10 + digit(pszDateTime[pos + 3]);
+
+	const vm::var<CellRtcDateTime> date_time{{ year, month, day, hour, minute, second, 0 }};
 
-	cellRtc.todo("heh year: %d, month: %d, day: %d, hour: %d, minute: %d, second: %d, tz: %s, tz_d: %d", t.tm_year + 1900, t.tm_mon + 1, t.tm_mday, t.tm_hour, t.tm_min, t.tm_sec, tz, timezoneMins);
-	date_time->year        = t.tm_year + 1900;
-	date_time->month       = t.tm_mon + 1;
-	date_time->day         = t.tm_mday;
-	date_time->hour        = t.tm_hour;
-	date_time->minute      = t.tm_min;
-	date_time->second      = t.tm_sec;
-	date_time->microsecond = 0;
 	cellRtcGetTick(date_time, pUtc);
-	cellRtcTickAddMinutes(pUtc, pUtc, timezoneMins);*/
 
 	return CELL_OK;
 }
 
 // Rfc3339: 1995-12-03T13:23:00.00Z
 error_code cellRtcParseRfc3339(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDateTime)
 {
-	cellRtc.notice("cellRtcParseRfc3339(pUtc=*0x%x, pszDateTime=%s)", pUtc, pszDateTime);
+	cellRtc.notice("cellRtcParseRfc3339(pUtc=*0x%x, pszDateTime=*0x%x)", pUtc, pszDateTime);
```

Comment:
Keep %s, why change this?

Reply:
Didn't realize it handles nullptr properly, do you want me to use %s for the other parsers as well then?